pkg/reconciler/apis/apiexportendpointslice/apiexportendpointslice_controller_test.go

```diff
@@ -0,0 +1,209 @@
+								ExternalURL: "https://server-1.kcp.dev/",
+							},
+						},
+						{
+							ObjectMeta: metav1.ObjectMeta{
+								Annotations: map[string]string{
+									logicalcluster.AnnotationKey: "root:org:ws",
+								},
+								Name: "shard2",
+							},
+							Spec: tenancyv1alpha1.ClusterWorkspaceShardSpec{
+								ExternalURL: "https://server-2.kcp.dev/",
+							},
+						},
+					}, nil
+				},
+				getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
+					if tc.apiExportMissing {
+						return nil, apierrors.NewNotFound(apisv1alpha1.Resource("APIExport"), name)
+					} else if tc.apiExportHasInvalidRef {
+						return nil, fmt.Errorf("internal error")
+					} else {
+						return &apisv1alpha1.APIExport{
+							ObjectMeta: metav1.ObjectMeta{
+								Annotations: map[string]string{
+									logicalcluster.AnnotationKey: "root:org:ws",
+								},
+								Name: "my-export",
+							},
+						}, nil
+					}
+				},
+			}
+
+			apiExportEndpointSlice := &apisv1alpha1.APIExportEndpointSlice{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{
+						logicalcluster.AnnotationKey: "root:org:ws",
+					},
+					Name: "my-slice",
+				},
+				Spec: apisv1alpha1.APIExportEndpointSliceSpec{
+					APIExport: apisv1alpha1.ExportReference{
+						Workspace: &apisv1alpha1.WorkspaceExportReference{
+							Path:       "root:org:ws",
+							ExportName: "my-export",
+						},
+					},
+				},
+			}
+			err := c.reconcile(context.Background(), apiExportEndpointSlice)
+			if tc.wantError {
+				require.Error(t, err, "expected an error")
+			} else {
+				require.NoError(t, err, "expected no error")
+			}
+
+			if tc.wantAPIExportEndpointSliceURLsError {
+				requireConditionMatches(t, apiExportEndpointSlice,
+					conditions.FalseCondition(
+						apisv1alpha1.APIExportEndpointSliceURLsReady,
+						apisv1alpha1.ErrorGeneratingURLsReason,
+						conditionsv1alpha1.ConditionSeverityError,
+						"",
+					),
+				)
+			}
+
+			if tc.wantAPIExportEndpointSliceURLsReady {
+				requireConditionMatches(t, apiExportEndpointSlice, conditions.TrueCondition(apisv1alpha1.APIExportEndpointSliceURLsReady))
+			}
+
+			if tc.wantAPIExportNotValid {
+				requireConditionMatches(t, apiExportEndpointSlice,
+					conditions.FalseCondition(
+						apisv1alpha1.APIExportValid,
+						tc.errorReason,
+						conditionsv1alpha1.ConditionSeverityError,
+						"",
+					),
+				)
+			}
+
+			if tc.wantAPIExportValid {
+				requireConditionMatches(t, apiExportEndpointSlice,
+					conditions.TrueCondition(apisv1alpha1.APIExportValid),
+				)
+			}
+		})
+	}
+}
+
+// requireConditionMatches looks for a condition matching c in g. Only fields that are set in c are compared (Type is
+// required, though). If c.Message is set, the test performed is contains rather than an exact match.
+func requireConditionMatches(t *testing.T, g conditions.Getter, c *conditionsv1alpha1.Condition) {
```

Comment:
You can implement a type-specific comparator for `cmp.Diff()` to be able to use that library for the entire check over conditions, which gets you very nice output, etc.

Reply:
Thanks for pointing that. I have made the change. Note that the previous approach is used for the tests of apiexport and apibinding reconciliation. The previous approach makes a "contain" comparison for "message", whereas I am now ignoring it, which is fine here. It would be possible to have a "contain" comparison with cmp by using a comparer, but was not needed.